src/mpid/ch4/src/ch4r_symheap.h

```diff
@@ -222,71 +235,99 @@ static inline int MPIDI_CH4I_allocate_symshm_segment(MPIR_Comm * shm_comm_ptr,
         /* try to map with specified symmetric address. */
         mpl_err =
             MPL_shm_fixed_seg_create_and_attach(*shm_segment_hdl_ptr, shm_segment_len, base_ptr, 0);
-        if (mpl_err == MPL_SHM_EINVAL)
-            *mapfail_flag = 1;
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            all_map_result = map_result = (mpl_err == MPL_SHM_EINVAL) ?
+                MPIDI_CH4I_SYMSHM_MAP_FAIL : MPIDI_CH4I_SYMSHM_OTHER_FAIL;
+            goto root_sync;
+        } else
+            mapped_flag = true;
 
-        /* handle mapping failure separately, caller may retry predefined times before giving up */
-        MPIR_ERR_CHKANDJUMP((mpl_err != MPL_SHM_EINVAL && mpl_err != MPL_SHM_SUCCESS),
-                            mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+        mpl_err = MPL_shm_hnd_get_serialized_by_ref(*shm_segment_hdl_ptr, &serialized_hnd_ptr);
+        if (mpl_err != MPL_SHM_SUCCESS)
+            all_map_result = map_result = MPIDI_CH4I_SYMSHM_OTHER_FAIL;
 
+      root_sync:
         /* broadcast the mapping result on rank 0 */
-        mpi_errno = MPIR_Bcast(mapfail_flag, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+        mpi_errno = MPIR_Bcast(&map_result, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
         MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
 
         /* broadcast handle to the others if mapping is successful */
-        if (!*mapfail_flag) {
-            mpl_err = MPL_shm_hnd_get_serialized_by_ref(*shm_segment_hdl_ptr, &serialized_hnd_ptr);
-            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
-                                "**alloc_shar_mem");
-
+        if (map_result == MPIDI_CH4I_SYMSHM_SUCCESS) {
```

Comment:
Again, I would prefer
```C
if (map_result != MPIDI_CH4I_SYMSHM_SUCCESS)
     goto map_fail;
...
```
for readability.

Reply:
fixed